### PR TITLE
More efadv pyverbs tests

### PR DESCRIPTION
**Pull request by galpress**
This series contains a small fix to skip SRQ tests on unsupported devices and adds more efadv functionality to the testing infra.

### CODE REVIEW

File: tests/test_efadv.py [hunk 1]
```diff
@@ -7,10 +7,13 @@
 import errno
 from pyverbs.addr import AHAttr
 from pyverbs.base import PyverbsRDMAError
+from pyverbs.cq import CQ
+import pyverbs.enums as e
 from pyverbs.pd import PD
 import pyverbs.providers.efa.efadv as efa
 from tests.base import PyverbsAPITestCase
 from tests.test_addr import get_global_route
+from tests.test_qp import get_qp_init_attr
```
Comment (Kamalheib) [inline tests/test_efadv.py:16]: I think get_global_route() and get_qp_init_attr() can be moved to "tests.utils".

File: tests/test_efadv.py [hunk 2]
```diff
@@ -52,3 +55,21 @@ def test_efadv_query_ah(self):
                 if ex.error_code == errno.EOPNOTSUPP:
                     raise unittest.SkipTest('Not supported on non EFA devices')
                 raise ex
+
+
+class EfaQPTest(PyverbsAPITestCase):
+    """
+    Test SRD QP class
+    """
+    def test_efadv_create_driver_qp(self):
```
Comment (Kamalheib) [inline tests/test_efadv.py:64]: please add docstring

File: pyverbs/providers/efa/efadv.pxd [hunk 3]
```diff
@@ -15,3 +16,11 @@ cdef class EfaContext(Context):
...
 cdef class EfaDVDeviceAttr(PyverbsObject):
     cdef dv.efadv_device_attr dv
+
+
+cdef class EfaAH(AH):
+    pass
+
+
+cdef class EfaDVAHAttr(PyverbsObject):
+    cdef dv.efadv_ah_attr ah_attr
```
Comment (Kamalheib) [inline pyverbs/providers/efa/efadv.pxd]: No newline at end of the file.

File: pyverbs/providers/efa/efadv.pxd [hunk 4]
```diff
@@ -23,4 +24,8 @@ cdef class EfaAH(AH):
 
 
 cdef class EfaDVAHAttr(PyverbsObject):
-    cdef dv.efadv_ah_attr ah_attr
\ No newline at end of file
+    cdef dv.efadv_ah_attr ah_attr
+
+
+cdef class SRDQP(QP):
+    pass
```
Comment (Kamalheib) [inline pyverbs/providers/efa/efadv.pxd:31]: No newline at end of the file.

File: tests/test_efadv.py [hunk 5]
```diff
@@ -5,9 +5,12 @@
 """
 
 import errno
+from pyverbs.addr import AHAttr
 from pyverbs.base import PyverbsRDMAError
+from pyverbs.pd import PD
 import pyverbs.providers.efa.efadv as efa
 from tests.base import PyverbsAPITestCase
+from tests.test_addr import get_global_route
```
Comment (Kamalheib) [inline tests/test_efadv.py:15]: I think get_global_route() can be moved to "tests.utils".

File: tests/test_efadv.py [hunk 6]
```diff
@@ -29,3 +32,23 @@ def test_efadv_query(self):
                     if ex.error_code == errno.EOPNOTSUPP:
                         raise unittest.SkipTest('Not supported on non EFA devices')
                     raise ex
+
+
+class EfaAHTest(PyverbsAPITestCase):
+    """
+    Test functionality of the EfaAH class
+    """
+    def test_efadv_query_ah(self):
```
Comment (Kamalheib) [inline tests/test_efadv.py:44]: please add docstring

**Review by Kamalheib — CHANGES_REQUESTED** [review]

File: tests/test_efadv.py (hunk 1 above)
Comment (galpress) [reply tests/test_efadv.py:16]: Ack

File: tests/test_efadv.py (hunk 2 above)
Comment (galpress) [reply tests/test_efadv.py:64]: Ack

File: pyverbs/providers/efa/efadv.pxd (hunk 3 above)
Comment (galpress) [reply pyverbs/providers/efa/efadv.pxd]: Ack

File: pyverbs/providers/efa/efadv.pxd (hunk 4 above)
Comment (galpress) [reply pyverbs/providers/efa/efadv.pxd:31]: Ack

File: tests/test_efadv.py (hunk 5 above)
Comment (galpress) [reply tests/test_efadv.py:15]: Ack

File: tests/test_efadv.py (hunk 6 above)
Comment (galpress) [reply tests/test_efadv.py:44]: Ack

File: pyverbs/providers/efa/libefa.pxd [hunk 7]
```diff
@@ -17,5 +17,14 @@ cdef extern from 'infiniband/efadv.h':
         unsigned int            device_caps
         unsigned int            max_rdma_size
 
+    cdef struct efadv_ah_attr:
+        unsigned long comp_mask
+        unsigned short ahn
+        unsigned char reserved[6]
```
Comment (EdwardSro) [inline pyverbs/providers/efa/libefa.pxd:23]: It's preferred if you use  uint64_t, uint16_t etc. as they're defined in C (you can import them from libc.stdint)
But anyway that shouldn't matter as eventually they're linked to the C code by Cython...

File: pyverbs/providers/efa/libefa.pxd [hunk 8]
```diff
@@ -17,5 +17,14 @@ cdef extern from 'infiniband/efadv.h':
...
+
     int efadv_query_device(v.ibv_context *ibvctx, efadv_device_attr *attrs,
                            unsigned int inlen)
+    int efadv_query_ah(v.ibv_ah *ibvah, efadv_ah_attr *attr,
+                       unsigned int inlen)
+    v.ibv_qp *efadv_create_driver_qp(v.ibv_pd *ibvpd, v.ibv_qp_init_attr *attr,
+                                     unsigned driver_qp_type)
```
Comment (EdwardSro) [inline pyverbs/providers/efa/libefa.pxd:30]: Same thing here (uint32_t)

**Review by EdwardSro — COMMENTED** [review]
Other than the following and Kamal's comments, it looks ok to me.
Thanks Kamal.

File: pyverbs/providers/efa/libefa.pxd (hunk 7 above)
Comment (galpress) [reply pyverbs/providers/efa/libefa.pxd:23]: Ack

File: pyverbs/providers/efa/libefa.pxd (hunk 8 above)
Comment (galpress) [reply pyverbs/providers/efa/libefa.pxd:30]: Ack

**Review by EdwardSro — APPROVED** [review]

**Pull request merged**